src/sync-utils/sync-device.ts

```diff
@@ -0,0 +1,187 @@
+import fs from 'fs-extra';
+import {
+  DlpAddSyncLogEntryReqType,
+} from '../protocols/dlp-commands';
+import {DlpConnection} from '../protocols/sync-connections';
+import {readDbList} from './read-db';
+import {RestoreResourcesConduit} from '../conduits/restore-resources-conduit';
+import {SyncDatabasesConduit} from '../conduits/sync-databases-conduit';
+import {DownloadNewResourcesConduit} from '../conduits/download-rsc-conduit';
+import {InstallNewResourcesConduit} from '../conduits/install-rsc-conduit';
+import {UpdateClockConduit} from '../conduits/update-clock-conduit';
+import crypto from 'crypto';
+import debug from 'debug';
+import { UpdateSyncInfoConduit } from '../conduits/update-sync-info-conduit';
+import { ConduitData } from '../conduits/conduit-interface';
+
+const log = debug('palm-sync').extend('sync-device');
+
+const NO_ID_SET = 0;
+export const TO_INSTALL_DIR = 'install';
+export const DATABASES_STORAGE_DIR = 'databases';
+export const JSON_PALM_ID = 'palm-id.json';
+export const CARD_ZERO = 0;
+
+export class PalmDeviceLocalIdentification {
+  userId = 0;
+  userName = '';
+  newlySet = false;
+  thisPcId = 6789;
+}
+
+/**
+ * Set's how the conduits should behave when syncing
+ */
+export enum SyncType {
+  /** When a Palm already has an ID, and it's new to this PC */
+  FIRST_SYNC = 'FIRST SYNC',
+  /** When the last sync was not done on this PC */
+  SLOW_SYNC = 'SLOW SYNC',
+  /** When the last sync was done on this PC */
+  FAST_SYNC = 'FAST SYNC',
+}
+
+export async function syncDevice(
+  dlpConnection: DlpConnection,
+  storageDir: string,
+  requestedUserName: string
+) {
+  const palmDir = `${storageDir}/${requestedUserName}`;
+
+  let conduits = [
+    new SyncDatabasesConduit(),
+    new DownloadNewResourcesConduit(),
+    new InstallNewResourcesConduit(),
+    new UpdateClockConduit(),
+    new UpdateSyncInfoConduit()
+  ];
+
+  log(`Start syncing device! There are [${conduits.length}] conduits.`);
+  await assertMandatoryDirectiores(storageDir, palmDir);
+
+  let syncType = getDefaultSyncType();
+  let localID = getLocalID(dlpConnection, requestedUserName);
+
+  let shoudRestoreAllResources = false;
+
+  if (localID.userName != requestedUserName) {
+    throw new Error(
+      `Expected a Palm with user name [${requestedUserName}] but it is named [${localID.userName}] instead! `+
+      `Aborting sync!`
+    );
+  }
+
+  if (!fs.existsSync(`${palmDir}/${JSON_PALM_ID}`)) {
```

Comment:
Let's use `await fs.exists` and `await fs.writeJSON` here.

Reply:
Again, don't know why I've used them to begin with haha Done!